ambry-store/src/main/java/com/github/ambry/store/BootstrapController.java

```diff
@@ -0,0 +1,203 @@
+import com.github.ambry.clustermap.StateModelListenerType;
+import com.github.ambry.clustermap.StateTransitionException;
+import com.github.ambry.config.StoreConfig;
+import com.github.ambry.server.StoreManager;
+import java.io.IOException;
+import java.util.Map;
+import java.util.regex.Pattern;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.clustermap.StateTransitionException.TransitionErrorCode.*;
+
+
+public class BootstrapController {
+
+  private final String BOOTSTRAP_IN_PROGRESS_FILE_NAME;
+  private final String FILECOPY_IN_PROGRESS_FILE_NAME;
+  private final Pattern allLogSegmentFilesPattern = Pattern.compile("\\d+\\.log");
+  private final StoreConfig storeConfig;
+  private final StoreManager storeManager;
+  private final PartitionStateChangeListener storageManagerListener;
+  private final PartitionStateChangeListener fileCopyManagerListener;
+
+  private static final Logger logger = LoggerFactory.getLogger(BootstrapController.class);
+
+  public BootstrapController(
+      @Nonnull StoreManager storeManager,
+      @Nonnull StoreConfig storeConfig,
+      @Nonnull ClusterParticipant clusterParticipant) {
+    this.storeConfig = storeConfig;
+    this.storeManager = storeManager;
+
+    this.BOOTSTRAP_IN_PROGRESS_FILE_NAME = storeConfig.storeBootstrapInProgressFile;
+    this.FILECOPY_IN_PROGRESS_FILE_NAME = storeConfig.storeFileCopyInProgressFileName;
+
+    clusterParticipant.registerPartitionStateChangeListener(
+        StateModelListenerType.BootstrapControllerListener, new BootstrapControllerImpl());
+
+    Map<StateModelListenerType, PartitionStateChangeListener> partitionStateChangeListeners =
+        storeManager.getPrimaryClusterParticipant().getPartitionStateChangeListeners();
+
+    this.storageManagerListener =
+        partitionStateChangeListeners.get(StateModelListenerType.StorageManagerListener);
+
+    this.fileCopyManagerListener =
+        partitionStateChangeListeners.get(StateModelListenerType.FileCopyManagerListener);
+
+    logger.info("Bootstrap Controller's state change listener registered!");
+  }
+
+  public void start() {
+  }
+
+  class BootstrapControllerImpl implements PartitionStateChangeListener {
+
+    @Override
+    public void onPartitionBecomeBootstrapFromOffline(@Nonnull String partitionName) {
+      ReplicaId replica = storeManager.getReplica(partitionName);
+      PartitionStateChangeListener listenerToInvoke = null;
+
+      /**
+       * This algo is based on this design doc :-
+       * https://docs.google.com/document/d/1u57C0BU8oMMuMHC6Fh_B-6R612EaQb2AxtnBDlN6sAs
+       * The decision branches can be better understood with the doc.
+       */
+      if (null == replica) {
+        // there can be two scenarios:
+        // 1. this is the first time to add new replica onto current node;
+        // 2. last replica addition failed at some point before updating InstanceConfig in Helix
```

Comment:
This comment will not be relevant here, because we won't be doing any retries.

Reply:
+ We are not doing any clean ups at this level.

Reply:
Isn't it true that we'll see `replica = null` in two scenarios? :-
1. this is the first time to add new replica onto current node;
2. last replica addition failed at some point before updating InstanceConfig in Helix

Reply:
let's discuss this in more detail.

Reply:
just checked. you are right

Reply:
Removed the comment for now.